diff --git a/src/backend/egl/device.rs b/src/backend/egl/device.rs
--- a/src/backend/egl/device.rs
+++ b/src/backend/egl/device.rs
@@ -75,9 +75,11 @@ impl EGLDevice {
         if !extensions.iter().any(|s| s == "EGL_EXT_device_base") {
             return Err(Error::EglExtensionNotSupported(&["EGL_EXT_device_base"]));
         }

-        if !extensions.iter().any(|s| s == "EGL_EXT_device_enumeration") {
-            return Err(Error::EglExtensionNotSupported(&["EGL_EXT_device_enumeration"]));
+        if !extensions.iter().any(|s| s == "EGL_EXT_device_enumeration") {
+            warn!(
+                "Ignoring missing EGL_EXT_device_enumeration extension"
+            );
         }

         if !extensions.iter().any(|s| s == "EGL_EXT_device_query") {
             return Err(Error::EglExtensionNotSupported(&["EGL_EXT_device_query"]));
         }

diff --git a/src/backend/egl/display.rs b/src/backend/egl/display.rs
--- a/src/backend/egl/display.rs
+++ b/src/backend/egl/display.rs
@@ -194,13 +194,11 @@ unsafe fn select_platform_display<N: EGLNativeDisplay + 'static>(
-        let display = wrap_egl_call_ptr(|| unsafe {
-            ffi::egl::GetPlatformDisplayEXT(
-                platform.platform,
-                platform.native_display,
-                platform.attrib_list.as_ptr(),
-            )
-        })
-        .map_err(Error::DisplayCreationError);
+        let display = wrap_egl_call_ptr(|| unsafe {
+            ffi::egl::GetDisplay(
+                platform.native_display as *mut _
+            )
+        })
+        .map_err(Error::DisplayCreationError);

         let display = match display {
             Ok(display) => {
                 if display == ffi::egl::NO_DISPLAY {