From 6b69f5c248d85a15be4e08c0c4ed80fd7e35e557 Mon Sep 17 00:00:00 2001
From: Kajus Naujokaitis <kajusn@gmail.com>
Date: Fri, 13 Mar 2026 11:17:20 +0200
Subject: [PATCH] hide airplane mode toggle

Signed-off-by: Kajus Naujokaitis <kajusn@gmail.com>
---
 cosmic-applet-network/src/app.rs | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/cosmic-applet-network/src/app.rs b/cosmic-applet-network/src/app.rs
index 0c15d2bd..6bafb16a 100644
--- a/cosmic-applet-network/src/app.rs
+++ b/cosmic-applet-network/src/app.rs
@@ -1579,15 +1579,6 @@ impl cosmic::Application for CosmicNetworkApplet {
                 Element::from(
                     column![
                         vpn_ethernet_col,
-                        padded_control(
-                            toggler(self.nm_state.nm_state.airplane_mode,)
-                                .label(fl!("airplane-mode"))
-                                .on_toggle(Message::ToggleAirplaneMode)
-                                .text_size(14)
-                                .width(Length::Fill)
-                        ),
-                        padded_control(divider::horizontal::default())
-                            .padding([space_xxs, space_s]),
                     ]
                     .align_x(Alignment::Center)
                 ),
--
2.53.0

