From bd2074bea17243edd791c83f894f016a0cabb4b1 Mon Sep 17 00:00:00 2001
From: Kajus Naujokaitis <kajusn@gmail.com>
Date: Tue, 27 Jan 2026 08:50:12 +0200
Subject: [PATCH] network applet hide airplane mode toggle

Signed-off-by: Kajus Naujokaitis <kajus.naujokaitis@unikie.com>
---
 cosmic-applet-network/src/app.rs | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/cosmic-applet-network/src/app.rs b/cosmic-applet-network/src/app.rs
index ae956bd1..21118154 100644
--- a/cosmic-applet-network/src/app.rs
+++ b/cosmic-applet-network/src/app.rs
@@ -1587,20 +1587,6 @@ impl cosmic::Application for CosmicNetworkApplet {
                 Element::from(
                     column![
                         vpn_ethernet_col,
-                        padded_control(
-                            anim!(
-                                //toggler
-                                AIRPLANE_MODE,
-                                &self.timeline,
-                                fl!("airplane-mode"),
-                                self.nm_state.nm_state.airplane_mode,
-                                |_chain, enable| { Message::ToggleAirplaneMode(enable) },
-                            )
-                            .text_size(14)
-                            .width(Length::Fill)
-                        ),
-                        padded_control(divider::horizontal::default())
-                            .padding([space_xxs, space_s]),
                     ]
                     .align_x(Alignment::Center)
                 ),
-- 
2.52.0

