diff --git a/Makefile b/Makefile
index f0f6a4b..72d36b2 100644
--- a/Makefile
+++ b/Makefile
@@ -153,7 +153,7 @@ run: build/live.img $(EXT_FS) build/rootfs.verity.roothash
 	    -parallel none \
 	    -drive file=build/live.img,if=virtio,format=raw,readonly=on \
 	    -drive file=$(EXT_FS),if=virtio,format=raw,readonly=on \
-	    -append "console=ttyS0 roothash=$$(< build/rootfs.verity.roothash) ext=/dev/vdb intel_iommu=on" \
+	    -append "console=ttyLP0,115200 roothash=$$(< build/rootfs.verity.roothash) ext=/dev/vdb intel_iommu=on" \
 	    -device intel-iommu,intremap=on \
 	    -device virtio-vga-gl \
 	    -device vhost-vsock-pci,guest-cid=3
diff --git a/etc/s6-linux-init/run-image/service/getty-ttyLP0/run b/etc/s6-linux-init/run-image/service/getty-ttyLP0/run
new file mode 100755
index 0000000..b9f57a3
--- /dev/null
+++ b/etc/s6-linux-init/run-image/service/getty-ttyLP0/run
@@ -0,0 +1,6 @@
+#!/bin/execlineb -P
+# SPDX-License-Identifier: EUPL-1.2+
+# SPDX-FileCopyrightText: 2020-2021 Alyssa Ross <hi@alyssa.is>
+# SPDX-FileCopyrightText: 2022 Unikie
+
+getty -i -n -l /etc/login 0,115200,57600,38400,9600 ttyLP0 dumb
