From 13c0345fc3e5ed8b6c28d1f272bb2630881d4714 Mon Sep 17 00:00:00 2001
From: Vunny Sodhi <vunny.sodhi@unikie.com>
Date: Thu, 29 Aug 2024 19:46:24 +0300
Subject: [PATCH] x86/gpu: Don't reserve stolen memory for GPU passthrough

Signed-off-by: Vunny Sodhi <vunny.sodhi@unikie.com>
---
 arch/x86/kernel/early-quirks.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/x86/kernel/early-quirks.c b/arch/x86/kernel/early-quirks.c
index 59f4aefc6..3b735eea9 100644
--- a/arch/x86/kernel/early-quirks.c
+++ b/arch/x86/kernel/early-quirks.c
@@ -596,6 +596,9 @@ static void __init intel_graphics_quirks(int num, int slot, int func)
 	u16 device;
 	int i;
 
+	// Nothing to do for GPU passthrough case
+	return;
+
 	/*
 	 * Reserve "stolen memory" for an integrated GPU.  If we've already
 	 * found one, there's nothing to do for other (discrete) GPUs.
-- 
2.40.1

