diff --git a/nvidia/platform/t23x/p3768/kernel-dts/Makefile b/nvidia/platform/t23x/p3768/kernel-dts/Makefile
index f306119fe8a3..f7fcf6ac2013 100644
--- a/nvidia/platform/t23x/p3768/kernel-dts/Makefile
+++ b/nvidia/platform/t23x/p3768/kernel-dts/Makefile
@@ -13,6 +13,9 @@ dtb-$(BUILD_ENABLE) += tegra234-p3767-0000-p3509-a02.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3767-0000-as-p3767-0001-p3509-a02.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3767-0000-as-p3767-0003-p3509-a02.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3767-0000-as-p3767-0004-p3509-a02.dtb
+
+dtb-$(BUILD_ENABLE) += tegra234-p3767-host-passthrough.dtb
+
 dtb-$(BUILD_ENABLE) += tegra234-p3767-0001-p3509-a02.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3767-0003-p3509-a02.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3767-0004-p3509-a02.dtb
diff --git a/nvidia/platform/t23x/p3768/kernel-dts/tegra234-p3767-host-passthrough.dtb b/nvidia/platform/t23x/p3768/kernel-dts/tegra234-p3767-host-passthrough.dtb
new file mode 100644
index 000000000000..da5f0ec048a0
--- /dev/null
+++ b/nvidia/platform/t23x/p3768/kernel-dts/tegra234-p3767-host-passthrough.dtb
@@ -0,0 +1,12 @@
+/dts-v1/;
+#include "tegra234-p3767-0000-p3509-a02.dts"
+
+/*
+ * Update the pci-e ethernet to be accessible from vfio/guest
+ */
+&pcie_c1_rp {
+    interconnect-names = "dma-mem", "write";
+    /delete-property/ iommus;
+    /delete-property/ msi-parent;
+    /delete-property/ msi-map;
+};
