diff --git a/nvidia/soc/t23x/kernel-dts/tegra234-soc/tegra234-soc-uart.dtsi b/nvidia/soc/t23x/kernel-dts/tegra234-soc/tegra234-soc-uart.dtsi
index 503cb27..993baf9 100644
--- a/nvidia/soc/t23x/kernel-dts/tegra234-soc/tegra234-soc-uart.dtsi
+++ b/nvidia/soc/t23x/kernel-dts/tegra234-soc/tegra234-soc-uart.dtsi
@@ -185,3 +185,10 @@
                status = "disabled";
        };
 };
+
+
+&uarti {
+       compatible = "arm,dummy";
+       iommus = <&smmu_niso0 TEGRA_SID_NISO1_SMMU_TEST>;
+       status = "okay";
+};
