--- a/nvdisplay/src/nvidia-modeset/src/nvkms-evo3.c
+++ b/nvdisplay/src/nvidia-modeset/src/nvkms-evo3.c
@@ -4325,13 +4325,9 @@
         pSurfaceDesc = &pNIso->pSurfaceEvo->planes[0].surfaceDesc;
         offset = pNIso->offsetInWords / 4;
         ctrlVal = 0;
-        if (pHwState->completionNotifier.awaken) {
-            ctrlVal = FLD_SET_DRF(C37E, _SET_NOTIFIER_CONTROL, _MODE,
-                                  _WRITE_AWAKEN, ctrlVal);
-        } else {
-            ctrlVal = FLD_SET_DRF(C37E, _SET_NOTIFIER_CONTROL, _MODE,
-                                  _WRITE, ctrlVal);
-        }
+        /* R5 emits FLIP_OCCURRED only for WRITE_AWAKEN. */
+        ctrlVal = FLD_SET_DRF(C37E, _SET_NOTIFIER_CONTROL, _MODE,
+                              _WRITE_AWAKEN, ctrlVal);
 
         SetWinNotifierSurfaceAddressAndControl(pChannel,
             pSurfaceDesc, offset, ctrlVal);
