diff --git a/nvidia/platform/t23x/concord/kernel-dts/Makefile b/nvidia/platform/t23x/concord/kernel-dts/Makefile
index 8e040d61f72f..dace3ef9a86d 100644
--- a/nvidia/platform/t23x/concord/kernel-dts/Makefile
+++ b/nvidia/platform/t23x/concord/kernel-dts/Makefile
@@ -25,6 +25,9 @@ dtb-$(BUILD_ENABLE) += tegra234-p3701-0000-p3737-0000-kexec.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3701-0004-p3737-0000.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3701-0005-p3737-0000.dtb
 dtb-$(BUILD_ENABLE) += tegra234-p3701-0008-p3737-0000.dtb
+
+dtb-$(BUILD_ENABLE) += tegra234-p3701-host-passthrough.dtb
+
 dtbo-$(BUILD_ENABLE) += tegra234-p3737-overlay.dtbo
 dtbo-$(BUILD_ENABLE) += tegra234-p3737-audio-codec-rt5658.dtbo
 dtbo-$(BUILD_ENABLE) += tegra234-p3737-audio-codec-rt5640.dtbo
diff --git a/nvidia/platform/t23x/concord/kernel-dts/tegra234-p3701-host-passthrough.dts b/nvidia/platform/t23x/concord/kernel-dts/tegra234-p3701-host-passthrough.dts
new file mode 100644
index 000000000000..5e9a0885d318
--- /dev/null
+++ b/nvidia/platform/t23x/concord/kernel-dts/tegra234-p3701-host-passthrough.dts
@@ -0,0 +1,12 @@
+/dts-v1/;
+#include "tegra234-p3701-0000-p3737-0000.dts"
+
+/*
+ * Update the pci-e wifi to be accessible from vfio/guest
+ */
+&pcie_c1_rp {
+    interconnect-names = "dma-mem", "write";
+    /delete-property/ iommus;
+    /delete-property/ msi-parent;
+    /delete-property/ msi-map;
+};
