From 22d608f97fb776e94be60253e1e6096b257a23e9 Mon Sep 17 00:00:00 2001
From: Kajus Naujokaitis <kajus.naujokaitis@unikie.com>
Date: Fri, 31 Oct 2025 13:29:27 +0200
Subject: [PATCH] Preselect Ghaf themes

Signed-off-by: Kajus Naujokaitis <kajus.naujokaitis@unikie.com>
---
 src/page/appearance.rs | 23 +++++++----------------
 1 file changed, 7 insertions(+), 16 deletions(-)

diff --git a/src/page/appearance.rs b/src/page/appearance.rs
index 31ecb15..fa95166 100644
--- a/src/page/appearance.rs
+++ b/src/page/appearance.rs
@@ -81,20 +81,7 @@ impl Page {
             }
         };

-        let mut themes = vec![
-            Theme {
-                name: "COSMIC Dark".to_string(),
-                handle: dark_icon(),
-                builder: ThemeBuilder::dark(),
-                is_dark: true,
-            },
-            Theme {
-                name: "COSMIC Light".to_string(),
-                handle: light_icon(),
-                builder: ThemeBuilder::light(),
-                is_dark: false,
-            },
-        ];
+        let mut themes = vec![];

         #[cfg(feature = "nixos")]
         let themes_dir_path = "/run/current-system/sw/share/cosmic-themes/";
@@ -149,7 +136,7 @@ impl Page {
             themes.extend(extra_themes.into_iter());
         }

-        Self {
+        let mut page = Self {
             theme_mode_config,
             dark_theme_config: cosmic::cosmic_theme::Theme::dark_config().ok(),
             dark_theme_builder_config: ThemeBuilder::dark_config().ok(),
@@ -157,7 +144,11 @@ impl Page {
             light_theme_builder_config: ThemeBuilder::light_config().ok(),
             themes,
             selected: if ThemeMode::default().is_dark { 0 } else { 1 },
-        }
+        };
+
+        page.update(Message::Select(page.selected));
+
+        page
     }

     pub fn update(&mut self, message: Message) -> Task<page::Message> {
--
2.51.0
