From aa6fb0fe666b137188a7e407487748725a0627ea Mon Sep 17 00:00:00 2001
From: Kajus Naujokaitis <kajusn@gmail.com>
Date: Thu, 18 Jun 2026 10:35:53 +0300
Subject: [PATCH] Preselect Ghaf themes

Signed-off-by: Kajus Naujokaitis <kajusn@gmail.com>
---
 src/page/appearance.rs | 23 +++++++----------------
 1 file changed, 7 insertions(+), 16 deletions(-)

diff --git a/src/page/appearance.rs b/src/page/appearance.rs
index 016f43a..2cc655a 100644
--- a/src/page/appearance.rs
+++ b/src/page/appearance.rs
@@ -86,20 +86,7 @@ impl Page {
             }
         };
 
-        let mut themes = vec![
-            Theme {
-                name: "COSMIC Dark".to_string(),
-                handle: dark_icon(),
-                builder: ThemeBuilder::dark(),
-                is_dark: true,
-            },
-            Theme {
-                name: "COSMIC Light".to_string(),
-                handle: light_icon(),
-                builder: ThemeBuilder::light(),
-                is_dark: false,
-            },
-        ];
+        let mut themes = vec![];
 
         #[cfg(feature = "nixos")]
         let themes_dir_path = "/run/current-system/sw/share/cosmic-themes/";
@@ -154,7 +141,7 @@ impl Page {
             themes.extend(extra_themes);
         }
 
-        Self {
+        let mut page = Self {
             theme_mode_config,
             dark_theme_config: cosmic::cosmic_theme::Theme::dark_config().ok(),
             dark_theme_builder_config: ThemeBuilder::dark_config().ok(),
@@ -163,7 +150,11 @@ impl Page {
             themes,
             selected: if ThemeMode::default().is_dark { 0 } else { 1 },
             frosted_glass: false,
-        }
+        };
+
+        page.update(Message::Select(page.selected));
+
+        page
     }
 
     pub fn update(&mut self, message: Message) -> Task<page::Message> {
