From 410cea4877925b0cb2537e64f6be75d80df51c1a Mon Sep 17 00:00:00 2001
From: Kajus Naujokaitis <kajusn@gmail.com>
Date: Wed, 4 Feb 2026 13:23:41 +0200
Subject: [PATCH] dont use secret agent

Signed-off-by: Kajus Naujokaitis <kajusn@gmail.com>
---
 src/page/wifi.rs | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/page/wifi.rs b/src/page/wifi.rs
index 5b279d9..254f610 100644
--- a/src/page/wifi.rs
+++ b/src/page/wifi.rs
@@ -532,7 +532,7 @@ impl Page {
                             ssid,
                             ap.hw_address,
                             ap.network_type,
-                            self.secret_tx.clone(),
+                            None,
                         ));
                 }
             }
@@ -606,7 +606,7 @@ impl Page {
                                 identity,
                                 hw_address,
                                 password,
-                                secret_tx,
+                                secret_tx: None,
                             });
                         }
                     })
--
2.52.0

