From 3113a1d2904556741792cf553afa44b501d5b4fb Mon Sep 17 00:00:00 2001
From: Kajus Naujokaitis <kajusn@gmail.com>
Date: Fri, 13 Mar 2026 11:40:27 +0200
Subject: [PATCH] dont use secret agent

Signed-off-by: Kajus Naujokaitis <kajusn@gmail.com>
---
 src/page/wifi.rs | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/page/wifi.rs b/src/page/wifi.rs
index 8adc150..449dd38 100644
--- a/src/page/wifi.rs
+++ b/src/page/wifi.rs
@@ -522,7 +522,7 @@ impl Page {
                         .unbounded_send(network_manager::Request::SelectAccessPoint(
                             ssid,
                             ap.network_type,
-                            self.secret_tx.clone(),
+                            None,
                             self.active_device.as_ref().map(|d| d.interface.clone()),
                         ));
                 }
@@ -595,7 +595,7 @@ impl Page {
                                 ssid: ssid.to_string(),
                                 identity,
                                 password,
-                                secret_tx,
+                                secret_tx: None,
                                 interface,
                             });
                         }
--
2.53.0

