From d22127f0fd61bbeba8c12378b3c5b46cc3064d63 Mon Sep 17 00:00:00 2001
From: Zephyr Lykos <git@mochaa.ws>
Date: Sat, 22 Jun 2024 18:48:54 +0800
Subject: [PATCH] auth: guard against race condition with messages

---
 src/auth.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/auth.c b/src/auth.c
index a3a99b3..a5eb33d 100644
--- a/src/auth.c
+++ b/src/auth.c
@@ -148,15 +148,15 @@ enum pwcheck auth_pw_check(const char *s) {
 	size_t len;
 	ssize_t nread;
 	nread = read(err_pipe[PIPE_PARENT], &len, sizeof(size_t));
-	if(nread > 0) {
-		error_string = malloc(len+1);
+	if(nread > 0 && len <= PAM_MAX_MSG_SIZE) {
+		error_string = malloc(PAM_MAX_MSG_SIZE);
 		nread = read(err_pipe[PIPE_PARENT], error_string, len);
 		error_string[nread] = '\0';
 		return PW_ERROR;
 	}
 	nread = read(out_pipe[PIPE_PARENT], &len, sizeof(size_t));
-	if(nread > 0) {
-		message_string = malloc(len+1);
+	if(nread > 0 && len <= PAM_MAX_MSG_SIZE) {
+		message_string = malloc(PAM_MAX_MSG_SIZE);
 		nread = read(out_pipe[PIPE_PARENT], message_string, len);
 		message_string[nread] = '\0';
 		return PW_MESSAGE;
-- 
2.40.1

